src/igraph_eulerian.c

```diff
@@ -0,0 +1,660 @@
+        } else if (degout == degin + 1) {
+            outgoing_excess++;
+            if (outgoing_excess > 1) {
+                excess = 1;
+            } else {
+                *start_of_path = i;
+            }
+        } else { /* degin == degout + 1 */
+            incoming_excess++;
+            if (incoming_excess > 1) {
+                excess = 1;
+                /* this is the end of the potential path */
+            }
+        }
+
+        if (excess) {
+            *has_path = 0;
+            *has_cycle = 0;
+            igraph_vector_destroy(&in_degree);
+            igraph_vector_destroy(&out_degree);
+            IGRAPH_FINALLY_CLEAN(2);
+
+            return IGRAPH_SUCCESS;
+        }
+    }
+
+    IGRAPH_VECTOR_INIT_FINALLY(&csize_weak, 0);
+    IGRAPH_VECTOR_INIT_FINALLY(&check_for_self_loops, 0);
+    IGRAPH_VECTOR_INIT_FINALLY(&cluster_member, 0);
+
+    /* count non-singleton clusters */
+    IGRAPH_CHECK(igraph_clusters(graph, &cluster_member, &csize_weak, NULL, IGRAPH_WEAK));
+    cluster_count = 0;
+    vsize = igraph_vector_size(&csize_weak);
+    for (i = 0; i < vsize; i++) {
+        if (VECTOR(csize_weak)[i] > 1) {
+            cluster_count++;
+        } else {
+            long int vsize2 = igraph_vector_size(&cluster_member);
+            for (j = 0; j < vsize2; j++) {
+                if (VECTOR(cluster_member)[j] == i) {
+                    igraph_vector_push_back(&check_for_self_loops, j);
+                    break;
+                }
+            }
+        }
+    }
+
+    /* weak connectivity excluding singletons (which are taken care of later on) */
+    if (cluster_count > 1) {
+        res_weak = 0;
+    }
+
+    /* weak && out_exc = 0 && in_exc = 0 means strong connectivity */
+    if ((outgoing_excess == 0 && incoming_excess == 0) && (res_weak)) {
+        *has_path = 1;
+        *has_cycle = 1;
+
+        vsize = igraph_vector_size(&out_degree);
+        for (i = 0; i < vsize; i++) {
+            if ((((long int) VECTOR(out_degree)[i]) > 0) || (((long int) VECTOR(in_degree)[i]) > 0)) {
+                *start_of_path = i;
+                break;
+            }
+        }
+
+    /* strong connectivity is NOT required for just a path */
+    } else if ((outgoing_excess == 1 && incoming_excess == 1) && (res_weak)) {
+        *has_path = 1;
+        *has_cycle = 0;
+    } else {
+        *has_path = 0;
+        *has_cycle = 0;
+    }
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &il, IGRAPH_OUT));
+    IGRAPH_FINALLY(igraph_inclist_destroy, &il);
+
+    /* check if singleton self loops destroy the Eulerian path/cycle */
+    vsize = igraph_vector_size(&check_for_self_loops);
+    for (i = 0; i < vsize; i++) {
+        igraph_vector_int_t *incedges;
+        incedges = igraph_inclist_get(&il, VECTOR(check_for_self_loops)[i]);
+        long int nc = igraph_vector_int_size(incedges);
+        if (nc > 0) {
+            *has_path = 0;
+            *has_cycle = 0;
+        }
+    }
```

Comment:
This is basically checking the degrees. Do not construct an inclist to do this. The degrees have already been computed with `igraph_degee` earlier.

Reply:
I redesigned it using only degrees, good point